cdap-api/src/main/java/io/cdap/cdap/api/ServiceDiscoverer.java

```diff
@@ -24,6 +24,18 @@
  */
 public interface ServiceDiscoverer {
 
+  /**
+   * Discover the base URL for a Service of an Application running in a Namespace, relative to which Service endpoints
+   * can be accessed
+   *
+   * @param namespaceId   Namespace of the application
+   * @param applicationId Application name
+   * @param serviceId     Service name
+   * @return URL for the discovered service or null if the service is not found
+   */
+  @Nullable
+  URL getServiceURL(String namespaceId, String applicationId, String serviceId);
```

Comment:
This is a non backward compatible change and should be done with default implementation on this interface. Please open a new PR to fix it. @albertshau please help to follow up

Reply:
Don't we allow adding a new method to an interface? This cannot have a default implementation unless it's moved to RuntimeContext, which doesn't seem like the right thing to do

Reply:
was thinking it should use the current namespace as the default, but we can just use 'system', we'll send a follow up

Reply:
Adding new method breaks existing implementation and it is not source compatible

Reply:
Default implementation can be on the interface (java 8 feature), which should just throw unsupported operation exception in this particular case

Reply:
Though actually, any default implementation will behave incorrectly. Wouldn't it be better if implementations are forced to implement this?